### PR TITLE
feat(FR-2723): version-mismatch UX notice and view-latest banner

### PR DESCRIPTION
**Pull request by yomybaby**
Resolves FR-2723 (gh #7019)

## Summary

F6 stretch goals: version-mismatch UX banner + notice.

- "View latest version" banner shown on every non-latest version page (build-time inject)
- "Not in selected version" inline notice when version switcher falls back to index (client-side detect via sessionStorage flag set by the inline switcher script)
- Both dismissible per-session via `sessionStorage`
- i18n: en/ko/ja/th localized via `WEBSITE_LABELS`
- Air-gap safe; ~5 KB additional JS (well under 25 KB per-page budget)
- New asset only emitted in versioned mode — flat-mode builds are byte-identical to before

## Implementation

- `templates/assets/version-banner.js` (new) — handles dismissal + notice reveal at runtime; reads from `data-` attributes so the script is content-hashed once and shared across every language × every version
- `src/website-builder.ts` — adds `buildVersionBanner` + `buildVersionNotice`; extends version-switcher inline script to set `sessionStorage["docs.notice.notInVersion"]` before fallback navigation
- `src/website-generator.ts` — writes the hashed `version-banner.js` asset only when `loadedVersions.enabled`
- `src/config.ts` — extends `WEBSITE_LABELS` with 4 keys (`bannerViewLatest`, `bannerViewLatestLink`, `bannerDismiss`, `noticeNotInVersion`) × 4 languages
- `src/styles-web.ts` — `.docs-banner` (warning yellow) + `.docs-notice` (info blue) styles, ~80 LoC
- `ARCHITECTURE.md` — new "Version mismatch UX (FR-2723)" subsection under F6

## Verification

- [x] `bash scripts/verify.sh` → ALL PASS (relay, lint, format, tsc)
- [x] `pnpm --filter backend.ai-docs-toolkit run test` → 48/48 pass
- [x] Flat mode (no `versions:`) — verified `dist/web/en/quickstart.html` has zero `docs-banner` / `docs-notice` / `version-banner.js` references; assets dir contains only styles + search + og-default (no version-banner.js)
- [x] Versioned mode (temporary `versions:` block with 26.03 latest + 25.16) — `assets/version-banner.5fa82353.js` written; `25.16/en/admin-menu.html` has banner with link to `../../26.03/en/admin-menu.html`; `26.03/en/admin-menu.html` has NO banner; both have hidden notice with localized template attribute
- [x] All 4 languages (en/ko/ja/th) — banner body and notice template render correct localized strings
- [x] Version-switcher inline script writes `sessionStorage.setItem('docs.notice.notInVersion', target + '::' + slug)` before fallback navigation
- [x] `pnpm run pdf:en` → exit 0 (PDF code path untouched)
- [x] `pnpm run pdf:ko` → exit 0 (FR-2721 CJK PDF fix non-regression)
- [x] Temp `versions:` block reverted from `webui-docs/docs-toolkit.config.yaml` before commit

## Wall-clock impact

Versioned-mode build with 1 language took ~2.0s for 2 versions × 29 chapters (vs ~1.0s for flat-mode 1 version × 29 chapters); banner/notice injection is constant-time per page so the impact is negligible.

## Stack

- F5 in main → F6 #7011 → F2 #7012 → FR-2721 #7027 → FR-2722 #7028 → This PR

## Things NOT shipped

- No CDN URLs / runtime API calls
- No PDF code touched
- No `versions:` enabled in `webui-docs/docs-toolkit.config.yaml` (used as temp test only, reverted)
- `localStorage` not used — only `sessionStorage` per spec

### CODE REVIEW

**Comment by yomybaby** [comment]
## PR #7029 Review — FR-2723 version-mismatch UX (banner + notice)

**Verdict: APPROVE (no blocking findings).** Implementation is tight, well-commented, and correctly scoped to versioned-mode builds. All XSS attack surfaces are properly escaped; sessionStorage usage is safe; spec compliance is solid across all six acceptance dimensions.

### Severity summary

| Severity | Count |
|----------|-------|
| CRITICAL | 0 |
| HIGH | 0 |
| MEDIUM | 0 |
| LOW | 4 (all optional / nice-to-have) |

---

### Security — clean

| Surface | Verdict | Notes |
|---|---|---|
| `sessionStorage` payload | Safe | Two keys (`docs.banner.viewLatestDismissed:<cur>:<latest>`, `docs.notice.notInVersion`) carry only operator-validated version labels (`^[0-9]+\.[0-9]+$`, validated in `versions.ts`) and slugs (constrained to `[\p{L}\p{N}\s-]` by `slugify`). No PII, no user input, no auth tokens. All `try/catch` around storage access defends private-mode and quota errors. |
| Banner template XSS | Safe | `bannerViewLatest` / `bannerViewLatestLink` / `noticeNotInVersion` come from the build-time `WEBSITE_LABELS` constant (not user-supplied). `{version}` / `{latestVersion}` placeholders are substituted **before** `escapeHtml`, so even if a label contained `<` or `&`, the final attribute / text would still be escaped (`website-builder.ts:507-519`). The notice's client-side `body.textContent = template.replace(...)` is also DOM-safe (text node, not HTML) (`version-banner.js:107`). |
| `data-message-template` interpolation | Safe | Template is escaped via `escapeHtml` when emitted as an HTML attribute, then read back through `getAttribute(...)` (browser auto-unescapes) and substituted into `textContent` — the full round-trip never touches `innerHTML`. |
| URL construction for "view latest" link | Safe | `${prefix}${ver.latest}/${context.metadata.lang}/${ver.slug}.html` — version label is regex-validated, lang is from `book.config.yaml`, slug is `slugify`'d (no quotes, no spaces). The whole href is `escapeHtml`'d before emission (`website-builder.ts:518`). No injection vector. |
| Inline switcher script `dest` URL | Safe | Built from `dataset.*` values which round-trip through HTML attribute encoding. Same constraints as above; no `encodeURIComponent` is needed because the values cannot contain URL-significant characters. |
| `escapeHtml` does not escape `'` | Safe in this PR | All attributes use double quotes consistently. Worth noting for future maintainers but not exploitable here. |

### Spec compliance — full pass

| FR-2723 acceptance criterion | Status | Evidence |
|---|---|---|
| Flat mode: no banner/notice | ✅ | `buildVersionBanner` / `buildVersionNotice` early-return when `versionContext` is undefined; `buildVersionBannerScriptTag` skips when `assets.versionBanner` is undefined; the asset itself is only written when `loadedVersions.enabled` (`website-generator.ts:349`). PR description verifies byte-identical flat-mode output. |
| Latest version page: no banner | ✅ | `buildVersionBanner` returns `""` when `ver.current === ver.latest` (`website-builder.ts:491`). |
| Non-latest page: banner present, dismissible, persists per-session | ✅ | Banner rendered visible-by-default (good for no-JS); `version-banner.js` reads `sessionStorage[BANNER_DISMISS_KEY:<cur>:<latest>]` and hides on match; click handler writes `"1"`. Per-pair key correctly scopes dismissal so dismissing on 25.16 doesn't suppress on 25.10. |
| Slug-missing target → notice on index page | ✅ | Inline switcher writes flag before fallback navigation (`website-builder.ts:632-635`); `version-banner.js:setupNotice()` reads flag, verifies target == current, fills `{version}`, reveals notice. (Minor UX nit below.) |
| All 4 languages localized | ✅ | `WEBSITE_LABELS` extends en/ko/ja/th with the 4 new keys (`config.ts:247-303`). Translations look natural. |
| sessionStorage (NOT localStorage) | ✅ | Both keys use `sessionStorage` exclusively. No `localStorage` references in the new code. |
| Air-gap (no CDN, no fetches) | ✅ | Hashed asset is local; no `fetch`, no CDN URL, no external dependency. |
| Banner/notice JS small (~1-2 KB) | ⚠️ | Source is **4848 bytes** (≈4.7 KB unminified). PR description claims "~1-2 KB minified" but no minifier runs in this build pipeline — the asset ships at source size. Still well under the 25 KB per-page budget; the spec-language target was aspirational. Not a blocker. |

### Cross-cutting

- **PDF non-regression**: PR description verifies `pnpm run pdf:en` and `pnpm run pdf:ko` exit 0; no PDF code paths are touched. FR-2721 CJK PDF fix is preserved (no edits to `pdf-renderer*` or font config).
- **Verification harness**: PR description claims `bash scripts/verify.sh` passes. No `react/*` files changed — pre-existing TS env errors are not relevant.

### Code quality observations (LOW, optional)

1. **Dismissal key design is sound, but consider per-`current` only.** The `<current>:<latest>` scoping handles the `25.16 → 25.10` example in the comment, but `latest` rarely changes within a session (it's a build-time constant per deploy). The extra `:<latest>` suffix is defensible (it auto-invalidates dismissals when latest changes between deploys, e.g. 26.03 ships and the user's earlier 25.16-vs-25.30 dismissal stops applying), but worth a comment explaining that intent. The current comment only mentions the `25.16 → 25.10` case.
2. **Notice copy slightly inaccurate after meta-refresh.** `buildIndexPage` is a 0s meta-refresh to the first chapter, so the notice "...viewing the version index instead" actually surfaces on the first chapter page, not the literal `index.html`. Most users won't notice (the destination *is* the version's landing page in spirit), but consider rephrasing to "...redirected to the latest available chapter in version X" if the team wants strict accuracy. Not worth shipping a fix for in this PR.
3. **Stale-flag clearing on mismatch is aggressive.** `setupNotice()` removes the sessionStorage flag whenever `targetVersion !== currentVersion` (lines 96-99). If a user (a) clicks switcher to 25.10 (slug missing → flag set), (b) clicks browser-back before navigation completes, (c) clicks switcher to 25.10 again — the flag from step (a) is consumed by the in-progress navigation. Edge case, harmless (just no notice on first visit), but conceptually the flag could be left in place to retry. Current behavior is fine.
4. **No new unit tests for the new builders.** `buildVersionBanner` / `buildVersionNotice` / version-banner.js are uncovered by Jest. The PR description says "48/48 pass" — those are pre-existing tests. A small snapshot test for `buildVersionBanner({...})` returning empty string when `current === latest`, and HTML containing `data-current-version` / `data-latest-version` otherwise, would catch future regressions cheaply. Not a blocker (the assertion surface is small and visually verifiable).

### Things NOT flagged (per request)

- Pre-existing TS env errors in `react/*` — out of scope.

---

**Recommendation: approve and merge.** No security or correctness issues; LOW items are optional polish that can ship in a follow-up if desired.

**Comment by yomybaby** [comment]
### Merge activity

* **Apr 26, 8:39 AM UTC**: The merge label 'flow:merge-queue' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.com/merges?org=lablup&repo=backend.ai-webui) once it meets the requirements.
* **Apr 26, 8:55 AM UTC**: `yomybaby` added this pull request to the [Graphite merge queue](https://app.graphite.com/merges?org=lablup&repo=backend.ai-webui).
* **Apr 26, 8:57 AM UTC**: Merged by the [Graphite merge queue](https://app.graphite.com/merges?org=lablup&repo=backend.ai-webui).

**Comment by yomybaby** [comment]
* **#7029** <a href="https://app.graphite.com/github/pr/lablup/backend.ai-webui/7029?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/lablup/backend.ai-webui/7029?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#7028** <a href="https://app.graphite.com/github/pr/lablup/backend.ai-webui/7028?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#7027** <a href="https://app.graphite.com/github/pr/lablup/backend.ai-webui/7027?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#7012** <a href="https://app.graphite.com/github/pr/lablup/backend.ai-webui/7012?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#7011** <a href="https://app.graphite.com/github/pr/lablup/backend.ai-webui/7011?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#7006** <a href="https://app.graphite.com/github/pr/lablup/backend.ai-webui/7006?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#7008](https://github.com/lablup/backend.ai-webui/pull/7008) <a href="https://app.graphite.com/github/pr/lablup/backend.ai-webui/7008?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#7004** <a href="https://app.graphite.com/github/pr/lablup/backend.ai-webui/7004?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6988** <a href="https://app.graphite.com/github/pr/lablup/backend.ai-webui/6988?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>flow:merge-queue</kbd> - adds this PR to the back of the merge queue
- <kbd>flow:hotfix</kbd> - for urgent changes, fast-track this PR to the front of the merge queue

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.com/invite/github/lablup?ref=merge-queue-instructions-in-stack-comment&prId=7029).

<h2></h2>
<sub>

An organization admin has required the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: packages/backend.ai-docs-toolkit/templates/assets/version-banner.js [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+      sessionStorage.removeItem(NOTICE_FLAG_KEY);
+    } catch (_) {
+      // Best effort.
+    }
+
+    var dismissBtn = notice.querySelector(".docs-notice__dismiss");
+    if (dismissBtn) {
+      dismissBtn.addEventListener("click", function () {
+        notice.hidden = true;
+      });
```
[the remark this answers is not in the document]
Comment (yomybaby) [reply packages/backend.ai-docs-toolkit/templates/assets/version-banner.js:123]: Tracked in [FR-2725](https://lablup.atlassian.net/browse/FR-2725) / #7021 for follow-up — per-session notice dismissal via `sessionStorage` aligns with the F1+F6 picker integration scope where the picker, switcher, and notice share a common dismissal/state contract.

File: packages/backend.ai-docs-toolkit/src/website-builder.ts [hunk 2]
```diff
@@ -523,7 +627,13 @@ function buildVersionSwitcher(context: WebPageContext): string {
...
-    var dest = availability[target]
+    var hasSlug = availability[target] === true;
+    if (!hasSlug) {
+      try {
+        sessionStorage.setItem('docs.notice.notInVersion', target + '::' + slug);
+      } catch (_) {}
+    }
+    var dest = hasSlug
       ? prefix + target + '/' + lang + '/' + slug + '.html'
       : prefix + target + '/' + lang + '/index.html';
```
[the remark this answers is not in the document]
Comment (yomybaby) [reply packages/backend.ai-docs-toolkit/src/website-builder.ts:831]: Tracked in [FR-2725](https://lablup.atlassian.net/browse/FR-2725) / #7021 for follow-up — reconciling the version-switcher fallback target (index.html vs first chapter vs notice copy) belongs to the F1+F6 picker integration sweep where landing-page semantics across versioned and flat modes are unified.

File: packages/backend.ai-docs-toolkit/src/config.ts [hunk 3]
```diff
@@ -225,7 +225,25 @@ const DEFAULT_FIGURE_LABELS: Record<string, string> = {
...
+ *   - `bannerDismiss`        — aria-label / tooltip for the dismiss "×" button
+ *                              on both the banner and the notice.
+ *   - `noticeNotInVersion`   — body text shown when the version switcher
+ *                              fell back to the version index because the
+ *                              same slug doesn't exist there. Includes a
+ *                              `{version}` placeholder for the target minor.
+ *
+ * The placeholder syntax is plain `{name}` (NOT i18next `{{name}}`) because
+ * the substitution happens in our own ~1 KB inline JS, not via an i18n
+ * library. See `templates/assets/version-banner.js`.
```
[the remark this answers is not in the document]
Comment (yomybaby) [reply packages/backend.ai-docs-toolkit/src/config.ts:267]: Tracked in [FR-2724](https://lablup.atlassian.net/browse/FR-2724) / #7020 for follow-up — JSDoc cleanup to accurately reflect build-time vs client-side templating boundaries fits the omnibus polish sweep.

File: packages/backend.ai-docs-toolkit/templates/assets/version-banner.js [hunk 4]
```diff
@@ -0,0 +1,136 @@
+/* docs-toolkit version-mismatch UX (FR-2723).
+ *
+ * Two pieces of UI live here:
+ *
+ *   1. "View latest version" banner — injected by website-builder.ts on
+ *      every NON-latest version page. Hides itself when the user clicks
+ *      the dismiss button (per-session via `sessionStorage`). Has no
+ *      runtime API call: the destination URL and label substitutions
+ *      come from `data-` attributes on the banner element.
```
[the remark this answers is not in the document]
Comment (yomybaby) [reply packages/backend.ai-docs-toolkit/templates/assets/version-banner.js:9]: Tracked in [FR-2724](https://lablup.atlassian.net/browse/FR-2724) / #7020 for follow-up — header-comment alignment with the actual `data-*` attribute usage (dismissal scoping only, not URL/label generation) is good polish-sweep cleanup.

File: packages/backend.ai-docs-toolkit/src/website-builder.ts [hunk 5]
```diff
@@ -448,6 +456,95 @@ function buildCodeCopyScriptTag(assets: PageAssets, prefix: string): string {
...
+  const ver = context.versionContext;
+  if (!ver) return "";
+  const labels = WEBSITE_LABELS[context.metadata.lang] ?? WEBSITE_LABELS.en;
+  const messageTemplate = labels.noticeNotInVersion;
+  const dismissLabel = labels.bannerDismiss;
+  return `<div class="docs-notice docs-notice--not-in-version" role="status" hidden data-current-version="${escapeHtml(ver.current)}" data-message-template="${escapeHtml(messageTemplate)}">
+  <span class="docs-notice__body"></span>
+  <button type="button" class="docs-notice__dismiss" aria-label="${escapeHtml(dismissLabel)}" title="${escapeHtml(dismissLabel)}">&times;</button>
+</div>`;
+}
```
[the remark this answers is not in the document]
Comment (yomybaby) [reply packages/backend.ai-docs-toolkit/src/website-builder.ts:739]: Tracked in [FR-2724](https://lablup.atlassian.net/browse/FR-2724) / #7020 for follow-up — adding toolkit unit tests for flat-vs-versioned output (banner/notice markup, asset tag, sessionStorage flag contract) is a worthwhile regression guard for the polish sweep.
